codegen/files/rest/server.go

```diff
@@ -1,11 +1,587 @@
 package rest
 
 import (
+	"fmt"
+	"text/template"
+
 	"goa.design/goa.v2/codegen"
+	"goa.design/goa.v2/codegen/files"
+	"goa.design/goa.v2/design"
 	"goa.design/goa.v2/design/rest"
+	"net/http"
+	"sort"
+	"strings"
+)
+
+//todo: in encode headers and mappings
+
+const serverStructT = `{{ printf "%s lists the %s service endpoint HTTP handlers." .HandlersStruct .ServiceName | comment }}
+type {{ .HandlersStruct }} struct {
+	{{- range .ActionData }}
+	{{ .EndpointName }} http.Handler
+	{{- end }}
+}
+`
+
+const serverConstructorT = `{{ printf "%s instantiates HTTP handlers for all the %s service endpoints." .Constructor .ServiceName | comment }}
+func {{ .Constructor }}(
+	e *endpoints.{{ .ServiceName }},
+	dec rest.RequestDecoderFunc,
+	enc rest.ResponseEncoderFunc,
+	logger goa.Logger,
+) *{{ .HandlersStruct }} {
+	return &{{ .HandlersStruct }}{
+		{{- range .ActionData }}
+		{{ .EndpointName }}: {{ .Constructor }}(e.{{ .EndpointName }}, dec, enc, logger),
+		{{- end }}
+	}
+}
+`
+
+const serverMountT = `{{ printf "%s configures the mux to serve the %s endpoints." .MountHandlers .ServiceName | comment }}
+func {{ .MountHandlers }}(mux rest.ServeMux, h *{{ .HandlersStruct }}) {
+	{{- range .ActionData }}
+	{{ .MountHandler }}(mux, h.{{ .EndpointName }})
+	{{- end }}
+}
+`
+
+const serverHandlerT = `{{ printf "%s configures the mux to serve the \"%s\" service \"%s\" endpoint." .MountHandler .ServiceName .EndpointName | comment }}
+func {{ .MountHandler }}(mux rest.ServeMux, h http.Handler) {
+	{{- range .Routes }}
+	mux.Handle("{{ .Method }}", "{{ .Path }}", h)
+	{{- end }}
+}
+`
+
+const serverHandlerConstructorT = `{{ printf "%s creates a HTTP handler which loads the HTTP request and calls the \"%s\" service \"%s\" endpoint." .Constructor .ServiceName .EndpointName | comment }}
+{{ comment "The middleware is mounted so it executes after the request is loaded and thus may access the request state via the rest package ContextXXX functions."}}
+func {{ .Constructor }}(
+	endpoint goa.Endpoint,
+	dec rest.RequestDecoderFunc,
+	enc rest.ResponseEncoderFunc,
+	logger goa.Logger,
+) http.Handler {
+	var (
+		decodeRequest  = {{ .Decoder }}(dec)
+		encodeResponse = {{ .Encoder }}EncodeResponse(enc)
+		encodeError    = {{ if .HasErrors }}{{ .ErrorEncoder }}{{ else }}EncodeError{{ end }}(enc, logger)
+	)
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		payload, err := decodeRequest(r)
+		if err != nil {
+			encodeError(w, r, goa.ErrInvalid("request invalid: %s", err))
+			return
+		}
+
+		res, err := endpoint(r.Context(), payload)
+
+		if err != nil {
+			encodeError(w, r, err)
+			return
+		}
+		if err := encodeResponse(w, r, res); err != nil {
```

Comment:
Ok i will not render the encodeResponse if there are no responses defined. But i do not think we could check on design.Empty for the body because we can still set some headers. Otherwise we should check if body and header is empty.